statsmodels/gam/gam.py

```diff
@@ -0,0 +1,767 @@
+# -*- coding: utf-8 -*-
+"""
+Generalized Additive Models
+
+Author: Luca Puggini
+Author: Josef Perktold
+
+created on 08/07/2015
+"""
+
+from __future__ import division
+import numpy as np
+from scipy import optimize
+from statsmodels.discrete.discrete_model import Logit
+from statsmodels.genmod.generalized_linear_model import (GLM, GLMResults,
+    GLMResultsWrapper, lm, _check_convergence)
+import statsmodels.regression._tools as reg_tools
+from statsmodels.tools.sm_exceptions import PerfectSeparationError
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.base._penalized import PenalizedMixin
+from statsmodels.gam.gam_penalties import MultivariateGamPenalty
+from statsmodels.tools.linalg import matrix_sqrt
+
+
+class GLMGAMResults(GLMResults):
+    """Results class for generalized additive models, GAM.
+
+    This inherits from GLMResults.
+
+    Warning: not all inherited methods might take correctly account of the
+    penalization
+
+    GLMGAMResults inherits from GLMResults
+
+    Extra Attributes
+    ----------------
+    edf : list of effective degrees of freedom for each column of the
+        design matrix.
+
+    hat_matrix_diag : diagonal of hat matrix
+
+    gcv : generalized cross-validation criterion computed as
+        `gcv = scale / (1. - hat_matrix_trace / self.nobs)**2`
+
+    cv : cross-validation criterion computed as
+        cv = ((resid_pearson / (1. - hat_matrix_diag))**2).sum() / self.nobs
+
+    Notes
+    -----
+    status: experimental
+
+    """
+
+    def __init__(self, model, params, normalized_cov_params, scale, **kwds):
+
+        # this is a messy way to compute edf and update scale
+        # need several attributes to compute edf
+        self.model = model
+        self.params = params
+        self.normalized_cov_params = normalized_cov_params
+        self.scale = scale
+        edf = self.edf.sum()
+        self.df_model = edf - 1  # assume constant
+        # need to use nobs or wnobs attribute
+        self.df_resid = self.df_resid = self.model.endog.shape[0] - edf
+
```

Comment:
`self.df_resid = ` is here twice.

Personally I find this clearer as `df_resid = nobs - (df_model + k_const)` where `k_const` is an indicator for the `# assume constant` comment.  Very nearly all the `df_resid`s in the codebase follow this pattern.